### ISSUE TITLE
Emacs doesn't recognize the backtrace error location format

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#5044]
**Original bug ID:** 5044
**Reporter:** bluestorm
**Status:** acknowledged (set by @damiendoligez on 2011-06-01T20:52:21Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Category:** emacs mode
**Related to:** #4628 #5070
**Monitored by:** @gasche

## Bug description

Backtrace error locations are not printed in the standard format recognized by emacs. The backtrace format is the following :

  Raised at file "gUnify.ml", line 191, characters 24-64
  Called from file "list.ml", line 69, characters 12-15
  ...

The compiler usually produces an error message formatted like this :

  File "gUnify.ml", line 147, characters 2-9:
  ...

**Comment by vicuna** [comment]
**Comment author:** @dbuenzli

Which version are you using ? 

This should be fixed in 3.11 see ocaml/ocaml#4628

Daniel

**Comment by vicuna** [comment]
**Comment author:** bluestorm

You are right, it is fixed under 3.11. Thanks for your patch, and sorry for not searching mantis before reporting.

It's strange that we can't access the character positions of the error message. I don't understand the precise use of the error regexps internal to the Caml mode, but I had a look at emacs compilation-error-regexp-alist documentation, and I came with the following fix, wich works fine in my .emacs :

```elisp
  (defun caml-change-error-alist-for-backtraces ()
    "Hook to change the compilation-error-regexp-alist variable, to
     search the ocaml backtraces for error locations"
    (interactive)
    (progn
      (setq compilation-error-regexp-alist-alist
            (append  
             '((caml-backtrace
  "^ *\\(?:Raised at\\|Called from\\) file \\(\"?\\)\\([^,\" \n\t<>]+\\)\\1,\
   lines? \\([0-9]+\\)-?\\([0-9]+\\)?\\(?:$\\|,\
  \\(?: characters? \\([0-9]+\\)-?\\([0-9]+\\)?:?\\)?\\)"
                2 (3 . 4) (5 . 6)))
             compilation-error-regexp-alist-alist))
      (setq compilation-error-regexp-alist
            (append compilation-error-regexp-alist '(caml-backtrace)))))
  
  (add-hook 'caml-mode-hook 'caml-change-error-alist-for-backtraces)
```

(The regexp is mainly an adaptation of the 'caml' regexp preset in emacs compile.el)

I also have the following, to spot error positions in assert failures (Assert_failure("test.ml", 1, 11)) :

```elisp
  (defun caml-change-error-alist-for-assert-failure ()
    "Hook to change the compilation-error-regexp-alist variable, to
     search the assert failure messages for error locations"
    (interactive)
    (progn
      (setq compilation-error-regexp-alist-alist
            (append  
             '((caml-assert-failure
                "Assert_failure(\"\\([^,\" \n\t<>]+\\)\", \\([0-9]+\\), \\([0-9]+\\))"
                1 2 3))
             compilation-error-regexp-alist-alist))
      (setq compilation-error-regexp-alist
            (append compilation-error-regexp-alist '(caml-assert-failure)))))
   
  (add-hook 'caml-mode-hook 'caml-change-error-alist-for-assert-failure)
```

Would it be possible to integrate something similar in caml.el ?

**Comment by vicuna** [comment]
**Comment author:** @damiendoligez

In 4.01.0+dev, the backtrace works fine, including the character positions.

I'm turning this into a feature wish to make it work also for assert failures.
